src/sources/postgresql_metrics.rs

```diff
@@ -0,0 +1,836 @@
+                hosts: config.get_hosts().to_owned(),
+            });
+        }
+
+        let mut tags = BTreeMap::new();
+        tags.insert("endpoint".into(), config_to_endpoint(&config));
+        tags.insert(
+            "host".into(),
+            match &config.get_hosts()[0] {
+                Host::Tcp(host) => host.clone(),
+                #[cfg(unix)]
+                Host::Unix(path) => path.to_string_lossy().to_string(),
+            },
+        );
+
+        let mut this = Self {
+            config,
+            tls_config,
+            client: None,
+            namespace,
+            tags,
+            datname_filter,
+        };
+        this.build_client().await?;
+        Ok(this)
+    }
+
+    async fn build_client(&mut self) -> Result<(), BuildError> {
+        self.client = Some(match &self.tls_config {
+            Some(tls_config) => {
+                let mut builder =
+                    SslConnector::builder(SslMethod::tls_client()).context(TlsFailed)?;
+                builder
+                    .set_ca_file(tls_config.ca_file.clone())
+                    .context(TlsFailed)?;
+                let connector = MakeTlsConnector::new(builder.build());
+
+                let (client, connection) =
+                    self.config
+                        .connect(connector)
+                        .await
+                        .with_context(|| ConnectionFailed {
+                            endpoint: config_to_endpoint(&self.config),
+                        })?;
+                tokio::spawn(connection);
+                client
+            }
+            None => {
+                let (client, connection) =
+                    self.config
+                        .connect(NoTls)
+                        .await
+                        .with_context(|| ConnectionFailed {
+                            endpoint: config_to_endpoint(&self.config),
+                        })?;
+                tokio::spawn(connection);
+                client
+            }
+        });
+
+        let version_row = self
+            .get_client()
+            .query_one("SELECT version()", &[])
+            .await
+            .with_context(|| VersionFailed {
+                endpoint: config_to_endpoint(&self.config),
+            })?;
+        let version = version_row
+            .try_get::<&str, &str>("version")
+            .with_context(|| VersionFailed {
+                endpoint: config_to_endpoint(&self.config),
+            })?;
+        debug!(message = "Connected to server.", endpoint = %config_to_endpoint(&self.config), server_version = %version);
+
+        Ok(())
+    }
+
+    fn get_client(&self) -> &Client {
+        self.client.as_ref().unwrap()
+    }
+
+    async fn collect(&mut self) -> stream::BoxStream<'static, Metric> {
+        let build_client = match self.client {
+            Some(_) => Ok(()),
+            None => self.build_client().await,
```

Comment:
I think this would be a little simpler if we created the client on start-up rather than delaying to the first scrape. Then `self.client` doesn't need to be an `Option`. What do you think?

Reply:
I initially add connect on start-up so there was no `Option` but then moved to scrape because we need support reconnects.